pkg/controller/direct/tags/tagkey_controller.go

```diff
@@ -0,0 +1,353 @@
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting TagsTagKey %q: %w", fqn, err)
+	}
+
+	a.actual = actual
+	return true, nil
+}
+
+func (a *TagsTagKeyAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	// There no FQN until after creation (server generated id).
+
+	log := klog.FromContext(ctx)
+	log.V(0).Info("creating TagsTagKey")
+
+	req := &pb.CreateTagKeyRequest{
+		TagKey: direct.ProtoClone(a.desired),
+	}
+
+	op, err := a.tagKeysClient.CreateTagKey(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating TagsTagKey: %w", err)
+	}
+
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("waiting for creation of TagsTagKey: %w", err)
+	}
+	log.V(0).Info("created TagsTagKey", "name", created.GetName())
+
+	// For compatibility, we set spec.resourceID after creation because this is a server-generated-id resource that we are migrating from terraform/DCL.
+	// More info in docs/ai/server-generated-id.md
+	resourceID := strings.TrimPrefix(created.GetName(), "tagKeys/")
+	if err := createOp.SetSpecResourceID(ctx, resourceID); err != nil {
+		return err
+	}
+
+	return a.updateStatus(ctx, createOp, created)
+}
+
+func (a *TagsTagKeyAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+
+	fqn := a.id.String()
+
+	req := &pb.UpdateTagKeyRequest{
+		TagKey: direct.ProtoClone(a.desired),
+	}
+	req.TagKey.Name = fqn
+
+	diff, updateMask, err := a.changedFields(ctx)
+	if err != nil {
+		return fmt.Errorf("getting changed fields for TagsTagKey %q: %w", fqn, err)
+	}
+	req.UpdateMask = updateMask
+
+	structuredreporting.ReportDiff(ctx, diff)
+
+	latest := a.actual
+	if len(req.UpdateMask.Paths) != 0 {
+		log.V(0).Info("updating TagsTagKey", "name", fqn)
+
+		op, err := a.tagKeysClient.UpdateTagKey(ctx, req)
+		if err != nil {
+			return fmt.Errorf("updating TagsTagKey %q: %w", fqn, err)
+		}
+
+		updated, err := op.Wait(ctx)
+		if err != nil {
+			return fmt.Errorf("waiting for update of TagsTagKey %s: %w", fqn, err)
+		}
+		log.V(0).Info("updated TagsTagKey", "name", fqn)
+		latest = updated
+	}
+
+	return a.updateStatus(ctx, updateOp, latest)
+}
+
+func (a *TagsTagKeyAdapter) updateStatus(ctx context.Context, op directbase.Operation, latest *pb.TagKey) error {
+	status := &krm.TagsTagKeyStatus{}
+
+	// NOTYET: observedState
+	// {
+	// 	mapCtx := &direct.MapContext{}
+	// 	status.ObservedState = TagsTagKeyObservedState_v1alpha1_FromProto(mapCtx, latest)
+	// 	if mapCtx.Err() != nil {
+	// 		return mapCtx.Err()
+	// 	}
+	// }
+
+	status.ExternalRef = direct.PtrTo(latest.GetName())
+
+	// Legacy status fields
+	status.Name = direct.PtrTo(strings.TrimPrefix(latest.GetName(), "tagKeys/"))
+	status.NamespacedName = direct.PtrTo(latest.GetNamespacedName())
+	status.CreateTime = direct.PtrTo(latest.GetCreateTime().AsTime().Format("2006-01-02T15:04:05Z07:00"))
```

Comment:
Where is "2006-01-02T15:04:05Z07:00" coming from? This isn't a test, this doesn't seem like it belongs in the actual controller.